File: elixir/say/lib/say.ex
defmodule Say do
  @doc """
  Translate a positive integer into English.
  """
  @spec in_english(integer) :: {atom, String.t()}
  def in_english(number) when number < 0 or number > 999_999_999_999, do: {:error, "number is out of range"}
  def in_english(number), do: {:ok, spell_number(number)}

  defp spell_number(number) do
    number
    |> Integer.digits()
    |> Enum.reverse()
    |> Stream.chunk_every(3)
    |> Stream.zip([:one, :thousand, :million, :billion, :trillion])
    |> Stream.map(fn {digits, unit} -> {Enum.reverse(digits), unit} end)
    |> Stream.map(fn {digits, unit} -> {spell(digits), unit} end)
    |> Stream.map(fn {spelled, unit} -> format(spelled, unit) end)
    |> Stream.map(&String.trim/1)
    |> Stream.filter(&(&1 != ""))
    |> Enum.reverse()
    |> Enum.join(" ")
  end

  defp spell([0]), do: "zero"
  defp spell([1]), do: "one"
  defp spell([2]), do: "two"
  defp spell([3]), do: "three"
  defp spell([4]), do: "four"
  defp spell([5]), do: "five"
  defp spell([6]), do: "six"
  defp spell([7]), do: "seven"
  defp spell([8]), do: "eight"
  defp spell([9]), do: "nine"

  defp spell([0, 0]), do: ""
  defp spell([1, 0]), do: "ten"
  defp spell([2, 0]), do: "twenty"
  defp spell([3, 0]), do: "thirty"
  defp spell([4, 0]), do: "forty"
  defp spell([5, 0]), do: "fifty"
  defp spell([8, 0]), do: "eighty"
  defp spell([n, 0]), do: "#{spell([n])}ty"

  defp spell([1, 1]), do: "eleven"
  defp spell([1, 2]), do: "twelve"
  defp spell([1, 3]), do: "thirteen"
  defp spell([1, 5]), do: "fifteen"
  defp spell([1, n]), do: "#{spell([n])}teen"
  defp spell([m, n]), do: "#{spell([m, 0])}-#{spell([n])}"

  defp spell([0, 0, 0]), do: ""
  defp spell([0, 0, n]), do: spell([n])
  defp spell([0, m, n]), do: spell([m, n])
  defp spell([l, m, n]), do: "#{spell([l])} #{:hundred} #{spell([m, n])}"

  defp format(spelled_digits, :one), do: spelled_digits
  defp format("", _), do: ""
  defp format(spelled_digits, unit), do: "#{spelled_digits} #{unit}"

end

File: c/say/say.c
#include "say.h"
#include <stdio.h>
#include <stdlib.h>

static const char *repr_digit(int n) {
  switch (n) {
  case 0:
    return "zero";
  case 1:
    return "one";
  case 2:
    return "two";
  case 3:
    return "three";
  case 4:
    return "four";
  case 5:
    return "five";
  case 6:
    return "six";
  case 7:
    return "seven";
  case 8:
    return "eight";
  case 9:
    return "nine";
  default:
    return "";
  }
}

static const char *repr_prefix_20_to_99(int n) {
  switch (n / 10) {
  case 2:
    return "twenty";
  case 3:
    return "thirty";
  case 4:
    return "forty";
  case 5:
    return "fifty";
  case 6:
    return "sixty";
  case 7:
    return "seventy";
  case 8:
    return "eighty";
  case 9:
    return "ninety";
  default:
    return "";
  }
}

static const char *repr_10_to_19(int n) {
  switch (n % 10) {
  case 0:
    return "ten";
  case 1:
    return "eleven";
  case 2:
    return "twelve";
  case 3:
    return "thirteen";
  case 4:
    return "fourteen";
  case 5:
    return "fifteen";
  case 6:
    return "sixteen";
  case 7:
    return "seventeen";
  case 8:
    return "eighteen";
  case 9:
    return "nineteen";
  default:
    return "";
  }
}

static int write_0_to_99(int64_t n, char *buffer, int index) {
  if (n < 0 || n > 99) {
    return -1;
  }
  if (n < 10) {
    return sprintf(&buffer[index], "%s", repr_digit(n));
  } else if (n < 20) {
    return sprintf(&buffer[index], "%s", repr_10_to_19(n));
  } else {
    const char *prefix = repr_prefix_20_to_99(n);
    int d = n % 10;
    return (d == 0) ? sprintf(&buffer[index], "%s", prefix)
                    : sprintf(&buffer[index], "%s-%s", prefix, repr_digit(d));
  }
}

static int write_0_to_999(int64_t n, char *buffer, int index) {
  if (n < 0 || n > 999) {
    return -1;
  }
  if (n < 100) {
    return write_0_to_99(n, buffer, index);
  }
  const char *prefix = repr_digit(n / 100);
  int suffix = n % 100;
  if (suffix == 0) {
    return sprintf(&buffer[index], "%s hundred", prefix);
  } else {
    int prefix_length = sprintf(&buffer[index], "%s hundred ", prefix);
    int suffix_length = write_0_to_99(suffix, buffer, index + prefix_length);
    if (suffix_length == -1) {
      return -1;
    }
    return prefix_length + suffix_length;
  }
}

static int write_thousandths(int64_t *n, char *buffer, int index,
                             int64_t divisor, const char *suffix) {
  int64_t d = *n / divisor;
  *n %= divisor;
  if (d == 0) {
    return 0;
  }
  if (d > 1000) {
    return -1;
  }
  int written = write_0_to_999(d, buffer, index);
  index += written;
  const char *f = *n > 0 ? " %s " : " %s";
  if (divisor > 1) {
    written += sprintf(&buffer[index], f, suffix);
  }
  return written;
}

const int buffer_length =
    sizeof("nine hundred ninety-nine billion nine hundred ninety-nine million "
           "nine hundred ninety-nine thousand nine hundred ninety-nine");

int say(int64_t input, char **ans) {
  if (input < 0 || input > 999999999999) {
    return -1;
  }
  if (input == 0) {
    *ans = malloc(sizeof("zero"));
    sprintf(*ans, "zero");
    return 0;
  }
  *ans = malloc(buffer_length * sizeof(char));
  int index = 0;
  int written;
  written = write_thousandths(&input, *ans, index, 1000000000, "billion");
  if (written == -1) {
    return -1;
  }
  index += written;
  written = write_thousandths(&input, *ans, index, 1000000, "million");
  if (written == -1) {
    return -1;
  }
  index += written;
  written = write_thousandths(&input, *ans, index, 1000, "thousand");
  if (written == -1) {
    return -1;
  }
  index += written;
  written = write_thousandths(&input, *ans, index, 1, "");
  if (written == -1) {
    return -1;
  }
  return 0;
}
